packages/python/plotly/plotly/express/_core.py

```diff
@@ -1648,109 +1846,212 @@ def process_dataframe_hierarchy(args):
     """
     Build dataframe for sunburst, treemap, or icicle when the path argument is provided.
     """
-    df = args["data_frame"]
+    df: nw.DataFrame = args["data_frame"]
     path = args["path"][::-1]
     _check_dataframe_all_leaves(df[path[::-1]])
-    discrete_color = False
+    discrete_color = not _is_continuous(df, args["color"]) if args["color"] else False
+
+    df = df.lazy()
 
-    new_path = []
-    for col_name in path:
-        new_col_name = col_name + "_path_copy"
-        new_path.append(new_col_name)
-        df[new_col_name] = df[col_name]
+    new_path = [col_name + "_path_copy" for col_name in path]
+    df = df.with_columns(
+        **{
+            new_col_name: nw.col(col_name)
+            for new_col_name, col_name in zip(new_path, path)
+        }
+    )
     path = new_path
     # ------------ Define aggregation functions --------------------------------
-
-    def aggfunc_discrete(x):
-        uniques = x.unique()
-        if len(uniques) == 1:
-            return uniques[0]
-        else:
-            return "(?)"
-
     agg_f = {}
-    aggfunc_color = None
     if args["values"]:
         try:
-            df[args["values"]] = pd.to_numeric(df[args["values"]])
-        except ValueError:
+            if isinstance(args["values"], Sequence) and not isinstance(
+                args["values"], str
+            ):
+                df = df.with_columns(
+                    **{c: nw.col(c).cast(nw.Float64()) for c in args["values"]}
+                )
+            else:
+                df = df.with_columns(
+                    **{args["values"]: nw.col(args["values"]).cast(nw.Float64())}
+                )
+
+        except Exception:  # pandas, Polars and pyarrow exception types are different
             raise ValueError(
                 "Column `%s` of `df` could not be converted to a numerical data type."
                 % args["values"]
             )
 
-        if args["color"]:
-            if args["color"] == args["values"]:
-                new_value_col_name = args["values"] + "_sum"
-                df[new_value_col_name] = df[args["values"]]
-                args["values"] = new_value_col_name
+        if args["color"] and args["color"] == args["values"]:
+            new_value_col_name = args["values"] + "_sum"
+            df = df.with_columns(**{new_value_col_name: nw.col(args["values"])})
+            args["values"] = new_value_col_name
         count_colname = args["values"]
     else:
         # we need a count column for the first groupby and the weighted mean of color
         # trick to be sure the col name is unused: take the sum of existing names
+        columns = df.collect_schema().names()
```

Comment:
What is the difference between this and `columns = df.columns` ?

Reply:
Oh I see, is it because it's a lazy df?

Reply:
yes correct, for a polars lazyframe, `.columns` raises a user warning